### Explore raw merged data

In [1]:
import pandas as pd
import json

In [2]:
data = pd.read_parquet('amazon_fashion_merged_051624.parquet')

In [3]:
data.head()

main_category                                              title  \
0  AMAZON FASHION  CHUVORA 925 Sterling Silver Open Celtic Knot C...   
1  AMAZON FASHION  XX-Large Slip Stop Single Tread Slipper Socks ...   
2  AMAZON FASHION  Sterling Silver 3mm Round Cut CZ Tennis Bracel...   
3  AMAZON FASHION  VERO MONTE 4 Pairs Womens TRULY No Show Socks ...   
4  AMAZON FASHION  SA106 Womens Rhinestone Jewel Polarized Lens 6...   

   average_rating  rating_number  \
0             4.3            137   
1             4.3           3096   
2             3.3             18   
3             4.0             58   
4             3.9             34   

                                            features  \
0  [STAMPED 925 STERLING SILVER - This high quali...   
1  [Cozy Slipper Socks For Men; Bariatric Slipper...   
2                                                 []   
3                                     [Machine Wash]   
4  [Imported, Plastic frame, anti-reflective lens...   

                                         description  price  \
0  [This beautiful jewelry would be a great addit...  30.00   
1  [Anyone looking for the best in single tread s...  20.99   
2                                                 []    NaN   
3                                                 []    NaN   
4  [Men's anti-glare lens sunglasses 100% UVA & U...  12.95   

                                              images  \
0  [{'hi_res': 'https://m.media-amazon.com/images...   
1  [{'hi_res': 'https://m.media-amazon.com/images...   
2  [{'hi_res': 'https://m.media-amazon.com/images...   
3  [{'hi_res': 'https://m.media-amazon.com/images...   
4  [{'hi_res': 'https://m.media-amazon.com/images...   

                                              videos       store  ...  \
0  [{'title': 'Irish Locket Celtic Locket Necklac...     CHUVORA  ...   
1                                                 []    PrimeMed  ...   
2                                                 []       KEZEF  ...   
3                                                 []  VERO MONTE  ...   
4                                                 []       SA106  ...   

  bought_together                                             rating  \
0            None  [5.0, 5.0, 2.0, 5.0, 5.0, 3.0, 1.0, 5.0, 5.0, ...   
1            None  [5.0, 1.0, 5.0, 2.0, 2.0, 5.0, 5.0, 5.0, 2.0, ...   
2            None  [2.0, 1.0, 5.0, 5.0, 5.0, 1.0, 5.0, 5.0, 5.0, ...   
3            None  [1.0, 1.0, 5.0, 5.0, 5.0, 1.0, 3.0, 5.0, 5.0, ...   
4            None  [5.0, 3.0, 5.0, 5.0, 5.0, 3.0, 4.0, 3.0, 5.0, ...   

                                        title_review  \
0  [Pretty locket, Great gift for my wife, OK.......   
1  [A, Cheaply made. Single wearing, PROVIDES SAF...   
2  [Two Stars, One Star, A perfect gift., Subtle,...   
3  [Won’t buy again, Slips from the feet (update ...   
4  [I LOVE these glasses, Great glass, but create...   

                                                text  \
0  [I think this locket is really pretty. The ins...   
1  [Great, Very loosely made material. Stretched ...   
2  [One of the stones fell out within the first 2...   
3  [Crappy socks. Money wasted. Bought to wear wi...   
4  [I LOVE these glasses!  They fit perfectly ove...   

                                       images_review  \
0  [{'attachment_type': 'IMAGE', 'large_image_url...   
1  [{'attachment_type': 'IMAGE', 'large_image_url...   
2                                                 []   
3  [{'attachment_type': 'IMAGE', 'large_image_url...   
4  [{'attachment_type': 'IMAGE', 'large_image_url...   

                                                asin  \
0  [B00LOPVX74, B00LOPVX74, B00LOPVX74, B00LOPVX7...   
1  [B07B4JXK8D, B07B4JXK8D, B07B4JXK8D, B07B4JXK8...   
2  [B007ZSEQ4Q, B007ZSEQ4Q, B007ZSEQ4Q, B007ZSEQ4...   
3  [B07F2BTFS9, B07F2BTFS9, B07F2BTFS9, B07F2BTFS...   
4  [B00PKRFU4O, B00PKRFU4O, B00PKRFWJC, B076J2JZB...   

                                             user_id  \
0  [AGBFYI2DDIKXC5Y4FARTY

'main_category' - str

 'title' - str

 'average_rating' - float64

 'rating_number' - int64

 'features' - array

 'description' - array

 'price' - float64

 'images' - array

 'videos' - array

 'store' - str

 'categories' - array

 'details' - dict

 'parent_asin' - str

 'bought_together' - None

 'rating' - array

 'title_review' - array

 'text' - array

 'images_review' - array

 'asin' - array

 'user_id' - array

 'timestamp' - array

 'helpful_vote' - array
 
 'verified_purchase' - array

In [24]:
# Export 

# import pickle
# l = list(data['details'].iloc[0].keys())

# with open("product_DETAILS_keys", "wb") as fp:
#     pickle.dump(l, fp)

In [65]:
data['verified_purchase'].iloc[0]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True])

In [26]:
data['features'].iloc[0]

array(['STAMPED 925 STERLING SILVER - This high quality of engraved celtic knot locket and chain is made from the finest sterling silver as indicated with 925 metal stamp. 925 sterling silver is made from 92.5% silver and 7.5% copper. The copper is added to stabilize the silver so that it can hold its beautiful shape.',
       "NICKEL AND LEAD FREE - These beautiful pendant and necklace contain absolutely no nickel or lead, making them safe for people with nickel and lead allergies. Add this great of pendant and necklace to your jewelry collection and have peace of mind every time you wear them. Men or Women' Daily Jewelry and Gift for Your Important One.",
       'PENDANT NECKLACE - This celtic knot pendant necklace are only 0.94 x 1.29 inches long (height including the bail) Inside Measurement 0.62 x 0.62 inches , The pendant comes with the sterling silver snake chain 18 inches long, Spring ring clasp, You can wear this of endearing necklace with a pair of jeans or wear them with a f

See if we can make a proxy network variable using USER_ID

In [60]:
ids = []
for i in range(1000):
    ids += list(data['user_id'].iloc[i])

print(len(ids))
print(len(set(ids)))

31664
30092


<font color="salmon"> We see that from the first 1000 products, there ARE users who have left reviews for more than one product. But there are not many of them. But enough to perhaps make a graph out of it!

Check timestamp variable - this is a UNIX date-time object that can be converted to pythonic datetime variable

In [63]:
from datetime import datetime

# Get the current date and time
current_datetime = datetime.now()

# Print the current date and time
print("Current date and time:", current_datetime)

Current date and time: 2024-05-16 18:21:25.368579


In [64]:
type(current_datetime)

datetime.datetime

### Modify data - create new datafiles
* Change datatype
* Remove unneeded features
* Construct more useful aggregate measures
* Concatenate lists of strings
* Use verified purchases only
* Reformat & engineer more useful features from original ones

In [2]:
filepath = '/Users/ez/Downloads/Amazon_Fashion.jsonl'

# # Read some
# with open(filepath, 'r') as f:
#     reviews = []
#     # Read and parse the first 10,000 lines as JSON data
#     for _ in range(100000):
#         line = f.readline()
#         if line:
#             try:
#                 reviews.append(json.loads(line))
#             except json.JSONDecodeError:
#                 # Handle malformed JSON lines, if any
#                 pass

# Read all
with open(filepath, 'r') as f:
    reviews = []
    for line in f:
        try:
            reviews.append(json.loads(line))
        except ValueError:
            pass

len(reviews)

2500939

In [3]:
filepath = '/Users/ez/Downloads/meta_Amazon_Fashion.jsonl'

# # Read some
# with open(filepath, 'r') as f:
#     meta = []
#     # Read and parse the first 10,000 lines as JSON data
#     for _ in range(100000):
#         line = f.readline()
#         if line:
#             try:
#                 meta.append(json.loads(line))
#             except json.JSONDecodeError:
#                 # Handle malformed JSON lines, if any
#                 pass

# Read all
with open(filepath, 'r') as f:
    meta = []
    for line in f:
        try:
            meta.append(json.loads(line))
        except ValueError:
            pass

len(meta)

826108

In [4]:
meta = pd.DataFrame(meta)
reviews = pd.DataFrame(reviews)

#### Start revising features - separately 

In [5]:
# Drop features from META 
meta = meta.drop(columns=['main_category', 'videos', 'categories', 'bought_together'])

# Drop features from REVIEWS
reviews = reviews.drop(columns=['asin'])

In [6]:
# REVIEWS: Drop rows where VERIFIED_PURCHASE == False
print('length reviews before', len(reviews))
reviews = reviews[reviews['verified_purchase'] == True]
print('length reviews after', len(reviews))

length reviews before 2500939
length reviews after 2337702


In [7]:
# REVIEWS: Drop verified_rows col
reviews = reviews.drop(columns=['verified_purchase'])

In [9]:
# META: FEATURES (list of str)
## Merge all strings in list together
## Take out str from list 
## Return STR

def merge_convert_to_str(lst):
    """
    Transforms a list of strings to one long string object.
    """
    output = ''.join(lst)
    return output

In [10]:
meta['features'] = meta['features'].apply(lambda x: merge_convert_to_str(x))

In [11]:
# META: DESCRIPTION (list of str)
## Take out str from list
def take_str_out_lst(lst):
    """
    From a list containing a single string, converts object to string (by taking
    it out).
    """
    if len(lst) != 0:
        return lst[0]
    else:
        return None

In [12]:
meta['description'] = meta['description'].apply(lambda x: take_str_out_lst(x))

In [13]:
print('none count:', meta['description'].isna().sum())
print('number of products:', len(meta))
print('number of products that have descriptions:', (len(meta) - meta['description'].isna().sum()))

none count: 766819
number of products: 826108
number of products that have descriptions: 59289


<font color="salmon"> Only 59,289 products out of 826,000+ have descriptions.

In [14]:
meta['details'].iloc[150]

{'Is Discontinued By Manufacturer': 'No',
 'Product Dimensions': '12 x 9 x 5 inches; 12 Ounces',
 'Item model number': '833989 401',
 'Department': 'girls',
 'Date First Available': 'June 23, 2016'}

<font color="salmon"> We can use `details` feature to check for additional info like `Date First Available`, `Department`.

<br>

These could be used in in a graph. 

`Date First Available` - do reviews change over time

`Department` - could serve as nodes or edges, helps determine traffic to particular departments

In [15]:
# REVIEWS - TIMESTAMP (convert from unix to datetime.datetime)
from datetime import datetime

def unix_to_datetime(unix_timestamp_ms):
    """
    Converts a Unix timestamp to month-day-year format.
    """
    # Convert milliseconds to seconds
    timestamp_s = unix_timestamp_ms / 1000 

    # Retrieve full datetime object
    dt_object = datetime.utcfromtimestamp(timestamp_s).strftime('%Y-%m-%d %H:%M:%S')

    # Parse the string into a datetime object
    dt_object = datetime.strptime(dt_object, '%Y-%m-%d %H:%M:%S')

    # Format the datetime object as year-month-date
    year_month_date = dt_object.strftime('%Y-%m-%d') 

    return year_month_date  

In [16]:
reviews['timestamp'] = reviews['timestamp'].apply(lambda x: unix_to_datetime(x))

In [17]:
# REVIEWS: HELPFUL_VOTE 
## Create log-weighted text review
## The idea = repeat a string the log-weighted number of times to represent 

import math

def square_root_weights(vote):
    """
    Apply square root weighting to a list of votes.
    """
    weighted_vote = int(math.sqrt(vote))
    return weighted_vote

In [18]:
reviews['helpful_vote_weighted'] = reviews['helpful_vote'].apply(lambda x: square_root_weights(x))

In [19]:
# REVIEWS: Create weighted text
def duplicate_and_concatenate(text, num_duplications):
    """
    Duplicate the text and concatenate it with the original text,
    separating each new iteration of text with a space.
    """
    duplicated_text = ' '.join(text for _ in range(num_duplications))
    return text + ' ' + duplicated_text

In [20]:
reviews['text_weighted'] = reviews.apply(lambda row: 
                            duplicate_and_concatenate(row['text'], 
                            row['helpful_vote_weighted']), axis=1)

In [21]:
reviews.head()

rating                 title  \
0     5.0         Pretty locket   
1     5.0                     A   
2     2.0             Two Stars   
3     1.0       Won’t buy again   
4     5.0  I LOVE these glasses   

                                                text images parent_asin  \
0  I think this locket is really pretty. The insi...     []  B00LOPVX74   
1                                              Great     []  B07B4JXK8D   
2  One of the stones fell out within the first 2 ...     []  B007ZSEQ4Q   
3  Crappy socks. Money wasted. Bought to wear wit...     []  B07F2BTFS9   
4  I LOVE these glasses!  They fit perfectly over...     []  B00XESJTDE   

                        user_id   timestamp  helpful_vote  \
0  AGBFYI2DDIKXC5Y4FARTYDTQBMFQ  2020-01-09             3   
1  AFQLNQNQYFWQZPJQZS6V3NZU4QBQ  2020-12-20             0   
2  AHITBJSS7KYUBVZPX7M2WJCOIVKQ  2015-05-23             3   
3  AFVNEEPDEIH5SPUN5BWC6NKL3WNQ  2018-12-31             2   
4  AHSPLDNW5OOUK2PLH7GXLACFBZNQ  2015-08-13             0   

   helpful_vote_weighted                                      text_weighted  
0                      1  I think this locket is really pretty. The insi...  
1                      0                                             Great   
2                      1  One of the stones fell out within the first 2 ...  
3                      1  Crappy socks. Money wasted. Bought to wear wit...  
4                      0  I LOVE these glasses!  They fit perfectly over...

In [22]:
meta.head()

title  average_rating  \
0  YUEDGE 5 Pairs Men's Moisture Control Cushione...             4.6   
1  DouBCQ Women's Palazzo Lounge Wide Leg Casual ...             4.1   
2  Pastel by Vivienne Honey Vanilla Girls' Trapez...             4.3   
3                                   Mento Streamtail             2.0   
4  RONNOX Women's 3-Pairs Bright Colored Calf Com...             4.3   

   rating_number                                           features  \
0             16                                                      
1              7                     Drawstring closureMachine Wash   
2             11                       Zipper closureHand Wash Only   
3              1  Thermoplastic Rubber soleHigh Density Premium ...   
4           3032  Pull On closureSize Guide: "S" fits calf 10-12...   

                                         description  price  \
0                                               None    NaN   
1                                               None    NaN   
2                                               None    NaN   
3  Slip on the Women's Mento and you're ready to ...  29.81   
4  Ronnox Calf Sleeves - Allowing Your Body to Pe...  17.99   

                                              images               store  \
0  [{'thumb': 'https://m.media-amazon.com/images/...            GiveGift   
1  [{'thumb': 'https://m.media-amazon.com/images/...              DouBCQ   
2  [{'thumb': 'https://m.media-amazon.com/images/...  Pastel by Vivienne   
3  [{'thumb': 'https://m.media-amazon.com/images/...          Guy Harvey   
4  [{'thumb': 'https://m.media-amazon.com/images/...              RONNOX   

                                             details parent_asin  
0  {'Package Dimensions': '10.31 x 8.5 x 1.73 inc...  B08BHN9PK5  
1  {'Package Dimensions': '15 x 10.2 x 0.4 inches...  B08R39MRDW  
2  {'Is Discontinued By Manufacturer': 'No', 'Pac...  B077KJHCJ4  
3  {'Package Dimensions': '11.22 x 4.72 x 4.33 in...  B0811M2JG9  
4  {'Is Discontinued By Manufacturer': 'No', 'Pac...  B07SB2892S

"None" in `description` = Nonetype

"NaN" in `price` = float64

In [23]:
# EXPORT original datasets, separately

# meta.to_parquet('meta_modified_051624.parquet', index=False)
# reviews.to_parquet('reviews_modified_051624.parquet', index=False)

### MERGE AGAIN

* Do merge first
* Check DETAILS dictionaries
* Construct aggregate features

In [2]:
# Import data
# reviews = pd.read_parquet('reviews_modified_051624.parquet')
# meta = pd.read_parquet('meta_modified_051624.parquet')

In [3]:
# Drop original helpful_vote and helpful_vote_weighted before merge - don't need
reviews = reviews.drop(columns=['helpful_vote'])
reviews = reviews.drop(columns=['helpful_vote_weighted'])

In [4]:
print(reviews.shape)
print(meta.shape)

(2337702, 8)
(826108, 10)


#### DO MERGE

In [5]:
# Rename reviews title & images
reviews.rename(columns={"title": "title_review", "images": "images_review"}, inplace=True)

In [6]:
# Merge
merged_data = pd.merge(meta, reviews, left_on='parent_asin', right_on='parent_asin', how='right')

In [7]:
merged_data.shape # One row per REVIEW, still need to aggregate/group

(2337702, 17)

Use DASK to parallelize grouping and other aggregate tasks

In [11]:
import dask.dataframe as dd
merged_data_dask = dd.from_pandas(merged_data, npartitions=4)

In [12]:
# Prep for aggregation
## Drop PARENT_ASIN, otherwise duplicated
## 
reviews.drop(columns=['parent_asin'], inplace=True)

# Which columns do we want to aggregate?
aggregate_columns = list(reviews.columns)
aggregate_columns

['rating',
 'title_review',
 'text',
 'images_review',
 'user_id',
 'timestamp',
 'text_weighted']

In [13]:
# Perform groupby and aggregation of cols to list using Dask df
aggregated_reviews_dask = merged_data_dask.groupby('parent_asin')[aggregate_columns].agg(list).reset_index()

In [14]:
# Convert Dask DF back to Pandas DF
aggregated_reviews = aggregated_reviews_dask.compute()

In [15]:
aggregated_reviews.shape

(776976, 8)

In [68]:
# Save
# aggregated_reviews.to_parquet('aggregated_reviews_unclean_051624.parquet', index=False)

In [16]:
print(aggregated_reviews.shape) # should be equal to meta shape?
print(meta.shape)
print(reviews.shape)

(776976, 8)
(826108, 10)
(2337702, 7)


In [18]:
print(2500939-2337702)

163237


<font color="salmon"> There's some 40,000 rows missing. I think this is because after dropping the 163237 unverified reviews, many of these reviews were solitary reviews for 30,000+ products. Since there are no reviews for these products, they were dropped from the aggregated table.

In [19]:
# Calculate average_rating_reviewers
def average_of_list(lst):
    """
    Calculate the average of a list of values.
    """
    if lst:
        avg = sum(lst) / len(lst)
        return round(avg, 2)
    else: return None

In [20]:
# Apply the custom function to each row in the 'values' column
aggregated_reviews['avg_rating_reviewers'] = aggregated_reviews['rating'].apply(average_of_list)

In [21]:
# Calculate coefficient of variation
import numpy as np

def coeff_of_var(lst):
    """
    Calculate coefficient of variation to measure spread of rating values.
    """
    if lst:
        mean = np.mean(lst)
        std_dev = np.std(lst)
        coeff_var = std_dev / mean
        return round(coeff_var, 2)
    else: return None

In [22]:
aggregated_reviews['coefvar_rating_reviewers'] = aggregated_reviews['rating'].apply(coeff_of_var)

In [23]:
aggregated_reviews.head()

parent_asin                                             rating  \
0  B00LOPVX74  [5.0, 5.0, 2.0, 5.0, 5.0, 3.0, 1.0, 5.0, 5.0, ...   
1  B07B4JXK8D  [5.0, 1.0, 5.0, 2.0, 2.0, 5.0, 5.0, 5.0, 2.0, ...   
2  B007ZSEQ4Q  [2.0, 1.0, 5.0, 5.0, 5.0, 1.0, 5.0, 5.0, 5.0, ...   
3  B07F2BTFS9  [1.0, 1.0, 5.0, 5.0, 5.0, 1.0, 3.0, 5.0, 5.0, ...   
4  B00XESJTDE  [5.0, 3.0, 5.0, 5.0, 5.0, 3.0, 4.0, 3.0, 5.0, ...   

                                        title_review  \
0  [Pretty locket, Great gift for my wife, OK.......   
1  [A, Cheaply made. Single wearing, PROVIDES SAF...   
2  [Two Stars, One Star, A perfect gift., Subtle,...   
3  [Won’t buy again, Slips from the feet (update ...   
4  [I LOVE these glasses, Great glass, but create...   

                                                text  \
0  [I think this locket is really pretty. The ins...   
1  [Great, Very loosely made material. Stretched ...   
2  [One of the stones fell out within the first 2...   
3  [Crappy socks. Money wasted. Bought to wear wi...   
4  [I LOVE these glasses!  They fit perfectly ove...   

                                       images_review  \
0  [[], [], [], [], [], [], [], [], [], [], [], [...   
1  [[], [], [], [], [], [], [], [], [], [], [], [...   
2  [[], [], [], [], [], [], [], [], [], [], [], [...   
3  [[], [], [], [], [], [], [], [], [], [], [], [...   
4  [[], [], [], [], [], [], [], [], [], [], [], [...   

                                             user_id  \
0  [AGBFYI2DDIKXC5Y4FARTYDTQBMFQ, AGQIISDDHDYLF2P...   
1  [AFQLNQNQYFWQZPJQZS6V3NZU4QBQ, AE73VVEIEFN4NIM...   
2  [AHITBJSS7KYUBVZPX7M2WJCOIVKQ, AHUYRZV7YZIEHPE...   
3  [AFVNEEPDEIH5SPUN5BWC6NKL3WNQ, AH6MYAW6MC5CEA7...   
4  [AHSPLDNW5OOUK2PLH7GXLACFBZNQ, AH6PUGVRXGZ4YXX...   

                                           timestamp  \
0  [2020-01-09, 2019-09-28, 2018-11-29, 2017-01-0...   
1  [2020-12-20, 2022-08-03, 2021-01-19, 2020-05-0...   
2  [2015-05-23, 2016-09-15, 2014-04-27, 2013-10-1...   
3  [2018-12-31, 2019-07-30, 2019-02-04, 2019-10-2...   
4  [2015-08-13, 2016-10-09, 2018-09-10, 2020-02-0...   

                                       text_weighted  avg_rating_reviewers  \
0  [I think this locket is really pretty. The ins...                  4.25   
1  [Great , Very loosely made material. Stretched...                  3.92   
2  [One of the stones fell out within the first 2...                  3.25   
3  [Crappy socks. Money wasted. Bought to wear wi...                  4.10   
4  [I LOVE these glasses!  They fit perfectly ove...                  4.08   

   coefvar_rating_reviewers  
0                      0.29  
1                      0.38  
2                      0.58  
3                      0.37  
4                      0.30

In [24]:
# Drop original rating column
aggregated_reviews = aggregated_reviews.drop(columns=['rating'])

In [25]:
# Concatenate strings in list to string

# Use function from earlier:
def merge_convert_to_str(lst):
    """
    Transforms a list of strings to one long string object.
    """
    output = ' '.join(lst)
    return output

In [26]:
aggregated_reviews['title_review_agg'] = aggregated_reviews['title_review'].apply(lambda x: merge_convert_to_str(x))

In [ ]:
# Drop title_review
aggregated_reviews['title_review'].drop(columns=['title_review'], inplace=True)

In [35]:
# Aggregate TEXT feature too, using merge_convert_to_str
aggregated_reviews['text_agg'] = aggregated_reviews['text'].apply(lambda x: merge_convert_to_str(x))

In [40]:
# Drop
aggregated_reviews = aggregated_reviews.drop(columns=['text'])

In [43]:
# Aggregate weighted text
aggregated_reviews['text_weighted_agg'] = aggregated_reviews['text_weighted'].apply(lambda x: merge_convert_to_str(x))

# Drop og
aggregated_reviews = aggregated_reviews.drop(columns=['text_weighted'])

In [51]:
# Function to extract dictionaries from arrays - Written entirely by ChatGPT
def extract_dicts_from_array(arr):
    dicts = []
    for sub_arr in arr:
        if isinstance(sub_arr, np.ndarray) and sub_arr.dtype == object:
            for item in sub_arr:
                if isinstance(item, dict):
                    dicts.append(item)
    return dicts

In [52]:
# Apply func
aggregated_reviews['images_review_cln'] = aggregated_reviews['images_review'].apply(extract_dicts_from_array)

In [59]:
# Drop og images_review
aggregated_reviews = aggregated_reviews.drop(columns=['images_review'])

In [60]:
aggregated_reviews.head()

parent_asin                                   title_review_agg  \
0  B00LOPVX74  Pretty locket Great gift for my wife OK..... F...   
1  B07B4JXK8D  A Cheaply made. Single wearing PROVIDES SAFETY...   
2  B007ZSEQ4Q  Two Stars One Star A perfect gift. Subtle Five...   
3  B07F2BTFS9  Won’t buy again Slips from the feet (update gr...   
4  B00XESJTDE  I LOVE these glasses Great glass, but create a...   

                                             user_id  \
0  [AGBFYI2DDIKXC5Y4FARTYDTQBMFQ, AGQIISDDHDYLF2P...   
1  [AFQLNQNQYFWQZPJQZS6V3NZU4QBQ, AE73VVEIEFN4NIM...   
2  [AHITBJSS7KYUBVZPX7M2WJCOIVKQ, AHUYRZV7YZIEHPE...   
3  [AFVNEEPDEIH5SPUN5BWC6NKL3WNQ, AH6MYAW6MC5CEA7...   
4  [AHSPLDNW5OOUK2PLH7GXLACFBZNQ, AH6PUGVRXGZ4YXX...   

                                           timestamp  avg_rating_reviewers  \
0  [2020-01-09, 2019-09-28, 2018-11-29, 2017-01-0...                  4.25   
1  [2020-12-20, 2022-08-03, 2021-01-19, 2020-05-0...                  3.92   
2  [2015-05-23, 2016-09-15, 2014-04-27, 2013-10-1...                  3.25   
3  [2018-12-31, 2019-07-30, 2019-02-04, 2019-10-2...                  4.10   
4  [2015-08-13, 2016-10-09, 2018-09-10, 2020-02-0...                  4.08   

   coefvar_rating_reviewers  \
0                      0.29   
1                      0.38   
2                      0.58   
3                      0.37   
4                      0.30   

                                            text_agg  \
0  I think this locket is really pretty. The insi...   
1  Great Very loosely made material. Stretched ou...   
2  One of the stones fell out within the first 2 ...   
3  Crappy socks. Money wasted. Bought to wear wit...   
4  I LOVE these glasses!  They fit perfectly over...   

                                   text_weighted_agg  \
0  I think this locket is really pretty. The insi...   
1  Great  Very loosely made material. Stretched o...   
2  One of the stones fell out within the first 2 ...   
3  Crappy socks. Money wasted. Bought to wear wit...   
4  I LOVE these glasses!  They fit perfectly over...   

                                   images_review_cln  
0  [{'attachment_type': 'IMAGE', 'large_image_url...  
1  [{'attachment_type': 'IMAGE', 'large_image_url...  
2                                                 []  
3  [{'attachment_type': 'IMAGE', 'large_image_url...  
4  [{'attachment_type': 'IMAGE', 'large_image_url...

In [61]:
# EXPORT 
# aggregated_reviews.to_parquet('aggregated_reviews_CLEAN_051624.parquet', index=False)

### Add product info columns based on matching id

In [63]:
# Merge based on aggregated_reviews (which dropped some products)
final_df = pd.merge(meta, aggregated_reviews, on='parent_asin', how='right')

In [67]:
print(aggregated_reviews.shape)
print(final_df.shape)

(776976, 9)
(776976, 18)


Number of row matches, as expected. Great!

In [70]:
# Function to filter out key-value pairs where value is not None
def filter_non_none(d):
    return {k: v for k, v in d.items() if v is not None}

In [71]:
# Apply
final_df['details'] = final_df['details'].apply(filter_non_none)

In [75]:
# Check
# final_df['details'].iloc[157]

{'Date First Available': 'March 11, 2016',
 'Package Dimensions': '9.6 x 5.5 x 1.7 inches; 4.8 Ounces'}

In [84]:
final_df.dtypes

title                        object
average_rating              float64
rating_number                 int64
features                     object
description                  object
price                       float64
images                       object
store                        object
details                      object
parent_asin                  object
title_review_agg             object
user_id                      object
timestamp                    object
avg_rating_reviewers        float64
coefvar_rating_reviewers    float64
text_agg                     object
text_weighted_agg            object
images_review_cln            object
dtype: object

In [94]:
type(final_df['images_review_cln'].iloc[0])

list

In [95]:
final_df['images_review_cln'].iloc[0]

[{'attachment_type': 'IMAGE',
  'large_image_url': 'https://images-na.ssl-images-amazon.com/images/I/61ibpqlhYoL.jpg',
  'medium_image_url': 'https://images-na.ssl-images-amazon.com/images/I/61ibpqlhYoL._SL800_.jpg',
  'small_image_url': 'https://images-na.ssl-images-amazon.com/images/I/61ibpqlhYoL._SL256_.jpg'}]

Looks good!

In [79]:
# SAVE!
# final_df.to_parquet('amazon_fashion_clean_051624.parquet')